### ISSUE TITLE
"Extension 'ESLint' is configured as a formatter but it cannot format 'typescript'-files"

### ISSUE THREAD

**Issue by charles-allen** [microsoft/vscode-eslint#1355]
The ESLint plugin seems to have stopped formatting TypeScript files. I searched online, changelogs, and open issues. I cannot find any reference to this. I wondered if maybe we're supposed to migrate to "code actions on save"?

### Actual
_Format document_
![Screen Shot 2021-10-19 at 11 13 24](https://user-images.githubusercontent.com/31621095/137843018-6b5d4507-cf55-48ee-b303-811e164ccb92.png)

_Format with..._
![Screen Shot 2021-10-19 at 11 19 14](https://user-images.githubusercontent.com/31621095/137843252-a6f0d4d5-371b-44e9-adf2-dfbc1a3ee28c.png)

### Expected
- ESLint formats the document

### Steps
- Open a .ts file
- Mess up the indentation
- Format document (or format on save)

### Context
> ❯ npx eslint -v                     
> v6.8.0

ESLint: Last updated 10/6/2021, 19:30:21

![Screen Shot 2021-10-19 at 11 20 58](https://user-images.githubusercontent.com/31621095/137843909-46020bae-1886-4ad2-b55f-f0e4e3ea2e72.png)

**Comment by harrydoerr** [comment]
I was running into this error earlier, and I found that I could get the plugin to run again on save after editing my .eslintrc.js and reloading the window. I run into the error again if I right-click and select "Format Selection" as the photo shows below:

<img width="438" alt="Screen Shot 2021-10-19 at 11 44 52 AM" src="https://user-images.githubusercontent.com/39283395/137972070-18ea5706-97c4-4f4e-9278-dcbfa36c4db8.png">

This is a link to a repository for reproduction: https://github.com/harrydoerr/vscode-eslint-example

My VS Code settings are as follow:
```json
"eslint.alwaysShowStatus": true,
  "eslint.workingDirectories": [{ 
      "directory": "./", 
      "!cwd": true
      }],
  "[javascript]": {
    "editor.defaultFormatter": "dbaeumer.vscode-eslint"
  },
  "[typescript]": {
    "editor.defaultFormatter": "dbaeumer.vscode-eslint"
  },
    "eslint.validate": [
      "javascript",
      "typescript",
      "html",
  ],
  "eslint.format.enable": true,
  "editor.formatOnSave": true,
  "editor.defaultFormatter": "dbaeumer.vscode-eslint",
  "eslint.options": {
    "configFile": "./.eslintrc.js",
  },
  "eslint.nodePath": "./node_modules",
```

Thank you, @dbaeumer, for all the work you put into this wonderful extension!

**Comment by charles-allen** [comment]
@dbaeumer - My issue is with "Format Document". I saw it yesterday in 2 completely different repos (one of which I know was working the day before, and one that was brand new). I will try it again now, and share a repo if I can still reproduce. Possibly the VSCode version is also relevant.

**Comment by charles-allen** [comment]
I have found a way to replicate the issue reliably, but it will be very difficult to provide a minimal repo (and that might no longer be relevant):

- Open a reasonably large Vue project in VSCode
- Quickly mess up the formatting
- Run format (keyboard shortcut or command palette)

And the resolution:

- Wait 10 seconds

It seems that this issue only exists when VSCode/extensions are still be initialized. After a few moments, normal behavior returned. My assumption is that:

- I hit the issue immediately after setting up my new project, deps, etc (before it settled?! too soon after configuring ESLint?!)
- I instinctively restarted VSCode to give it a clean load, ironically recreating the issue
- I tested it by opening a (larger) project that I knew worked, but again presumably didn't wait long enough because I immediately saw the error I was expecting to see

2 days ago the issue seemed to persist for some time (minutes). Today the issue resolves itself within ~10 seconds. I suspect that 2 days ago my Macbook was under heavy load (migrating to Google Drive for Desktop) and that was impeding VSCode. So... perhaps everything is normal, and my system was just operating outside its typical performance range & outside my normal expectations.

**Comment by lxghtless** [comment]
> The ESLint extension has no support for formatting selections since ESLint itself (the npm module) has not API for it. This is the reason why you can't do it.

It stopped working for format on save as well without a selection. I checked with my team at work and the same thing happened to everyone. We run a mix of Macs and Linux distros. I wish I could share a project to help diagnose the issue, but we can’t do that. If I find some extra time I’ll try to setup a reasonably similar project to share an example. Maybe it’s a coincidence that it happened at the same time the extension updated.

**Comment by dbaeumer** [comment]
@charles-allen I recognized that you were talking about format, that is why I didn't close the issue. But @harrydoerr and @lxghtless  reported on format on selection.

Without some steps to reproduce it is extremally hard to tell what is going on. Does this reproduce with all extensions disabled except ESLint. It could be a problem with another extension saturating the extension host.

**Comment by SerhiyZheliznjak** [comment]
For me Eslint formatter stopped working after updating to 1.62, for all previously supported extensions .js(x), .ts(x).
Eslint server is crashing with: 

```
Eslint server is starting
Connection to server is erroring. Shutting down server.
Eslint server stopped.
Connection to server is erroring. Shutting down server.
Connection to server is erroring. Shutting down server.
```

**Comment by SerhiyZheliznjak** [comment]
For me it was occurred  right after update to 1.62, not sure what I changed but after at least one more restart it works

**Comment by edvinv** [comment]
Same here it works with Save (CTRL+S) but not working with Format document (CTRL+SHIFT+I).

**Comment by Jared0430** [comment]
I was getting this issue too. It turns out when I ran `npx eslint --init`, the .eslintrc.js file it generated had

```js
module.exports = {
  ...
  parserOptions: {
    ecmaVersion: 13
  },
  ...
}
```

If I ran `npx eslint index.js` manually it complained `Parsing error: Invalid ecmaVersion`

I fixed it by change ecmaVersion from 13 to 12

**Comment by dmtr343** [comment]
I got rid of this error for myself by adding this to the **settings.json**:
```
    "eslint.workingDirectories": [
        {
            "mode": "auto"
        }
    ],
```

**Comment by 1ike** [comment]
> In addition you can the always trigger format by pressing Ctrl+S even if the file is saved.

Yes, it work. I can't set `eslint` as the default formatter, but I don't use it in a such manner. I tried to make it work with save via `onFocusChange` and stopped after I saw a comment with `Ctrl+S`. I think that there may be problems with vscode itself, so for now I'll wait with `Ctrl+S` and husky.

**Comment by zeppelin** [comment]
For me, removing [Babel ES6/ES7](https://marketplace.visualstudio.com/items?itemName=dzannotti.vscode-babel-coloring) did the trick. VSCode refused to recoginze ESLint as a formatter for js/ts files, regardless the configuration/settings.

**Comment by felipe-gustavo** [comment]
Folks, I solved my issue with the plugin just running `yarn eslint <file>`, the CLI will show for you what is wrong with the plugin, in my case I switched the branch to another branch that requires a plugin that I haven't downloaded yet, so I downnloaded this plugin and restarted my VSCode and the plugin came back to working again. Apparently the vscode-eslint aren't treating and showing this errors that  accindentaly caused :,)

**Comment by lancedang64** [comment]
> Folks, I solved my issue with the plugin just running `yarn eslint <file>`, the CLI will show for you what is wrong with the plugin, in my case I switched the branch to another branch that requires a plugin that I haven't downloaded yet, so I downnloaded this plugin and restarted my VSCode and the plugin came back to working again. Apparently the vscode-eslint aren't treating and showing this errors that accindentaly caused :,)

Doing this, I was able to see that my eslint error was:
```
Error: Rules with suggestions must set the `meta.hasSuggestions` property to `true`.
Occured while lining {path-to-file-being-linted}
Rule: "react-hooks/exhaustive-deps"
```

My co-worker who had the same problem provided me with this solution:
in package.json, upgrade "eslint-plugin-react-hooks": 4.2.0 to 4.3.0, run yarn install, then eslint will start working again.
I then proceed to change it back and run yarn install again and the problem was fixed.

He said probably React fixed it in 4.3.0. Hope this helps anyone that came here!

**Comment by Rabbitzzc** [comment]
![BIi3I5VJnj](https://user-images.githubusercontent.com/26913511/162392876-1690c871-b9ca-4925-b9e2-889fdc4aee94.png)

**Comment by martinmckenna** [comment]
Having the same issue. I'm finding that when I "Format Selection," it only works when my `settings.json` looks like this:

```
{
    "editor.defaultFormatter": "dbaeumer.vscode-eslint",
}
```

but then gives me the error in the OP when I change it to this:

```
{
    "[typescript]": {
        "editor.defaultFormatter": "dbaeumer.vscode-eslint"
    },
}
```

**Comment by farmisen** [comment]
> > Folks, I solved my issue with the plugin just running `yarn eslint <file>`, the CLI will show for you what is wrong with the plugin

this was the solution for me. All my vscode settings where correct what I knew since my settings are synced from another machine where I didn't have this problem. Running `yarn eslint ...` told it could not find `.eslintrc` that I mistyped `.eslinrc`

**Comment by dbaeumer** [comment]
@JerryMacedoCastro can you format TypeScript-React files with ESLint using the terminal?

**Comment by danieljames-dj** [comment]
I'm also facing the same issue. But I can format the TypeScript React files using the following command:

```
npm run lint -- --fix <filename>.ts
```

**Comment by JerryMacedoCastro** [comment]
> @JerryMacedoCastro can you format TypeScript-React files with ESLint using the terminal?

No, I couldn't. But after some time I see the problem was my config file. It was pointing to a lib that I change.

**Comment by jamesarosen** [comment]
One oddity I've found: I only get "Extension 'eslint' is configured as formatted but it cannot be used for 'JavaScript'-files" for files in `.github/`.

format-on-save, Format, and Format With… all work for me on `app/some.{js,json,jsx,ts,tsx}`. But if I try to run them on `.github/test.{js,json}`, I get the error and no formatting.

I've made sure that the file is not ignored in my `.gitignore` or `.eslintignore`. When I run `node_modules/.bin/eslint --fix .github/test.js` it works as expected.

**Comment by sebkasanzew** [comment]
I'm having this issue now a few days. I can format everything using the CLI and even the "Format Document" menu entry and the "format on save".

But the Keyboard Shortcut for "Format Document" shows the message from @JerryMacedoCastro 

> I'm having the same issue. ![image](https://user-images.githubusercontent.com/39466788/188501341-4faeb364-5bd6-443f-accd-2cee1faff80c.png) ![image](https://user-images.githubusercontent.com/39466788/188501361-b1505db4-9056-4b95-914c-02efbd8f5654.png)

**Comment by LittleHoopoe** [comment]
this sounds crazy... but I've tested everything, to get the ESLint configured as formatter again. In the end after updated every eslint dep eslint-react and other eslint-plugins, also restarted vscode ESLint is working again as formatter... I hope this helps someone

**Comment by 1ike** [comment]
Check your settings (global or for the workspase). The min set for the "format on save" must include:
```
  "eslint.format.enable": true,
  "editor.formatOnSave": true,
  "editor.defaultFormatter": "dbaeumer.vscode-eslint",
```

You can specify the last two fields for a specific language:
```
"[typescriptreact]": {
  ...settings (formatOnSave and/or defaultFormatter)
 },
```

Specify autosave if you want with `"files.autoSave": "onFocusChange",`.


Never more trendy `editor.codeActionsOnSave`! :)

**Comment by WhiteR4bbitt** [comment]
I was having this issue and it was also related to https://github.com/microsoft/vscode-eslint/issues/1462

I was able to get it fixed without any system restarts.

I had `eslint` installed in the package and globally.
I removed the global installation as well as the legacy typescript and current parser plugins.
I removed and freshly installed my `pnpm` packages.
Restarted VS Code and works!

I also did not notice any eslint config files mentioned in this thread, but making sure the PARSER is set-up is key and seems to vary by project.

This is my config file if it is relevant to anyone else with this issue popping up:
```javascript
// .eslintrc.js
module.exports = {
  root: true,
  env: {
    browser: true,
    node: true,
    es2022: true
  },
  extends: [
    '@nuxtjs/eslint-config-typescript',
    'standard-with-typescript'
  ],
  parserOptions: {
    sourceType: 'module',
    tsconfigRootDir: __dirname,
    project: ['./tsconfig.json']
  },
  ignorePatterns: [
    '**/node_modules/**',
    '**/dist/**',
    '*.config.(js|ts)',
    '.eslintrc.js',
  ],
  rules: { ...
  }
}
```

**Comment by dhianica** [comment]
i try changed name file eslintrc.js to eslintrc.cjs and its work fine

![image](https://user-images.githubusercontent.com/21213646/212588892-cb931d52-d8d0-4020-8ec2-3ba724b85b91.png)

**Comment by ikerlorente11** [comment]
> ![BIi3I5VJnj](https://user-images.githubusercontent.com/26913511/162392876-1690c871-b9ca-4925-b9e2-889fdc4aee94.png)

This make the solution for me

**Comment by Serpentarius13** [comment]
changing .eslintrc.json to .eslintrc does the trick

**Comment by danielwuachin** [comment]
its 17-05-2024 and still working, thanks! I have the same problem using eslint with prettier, here my config files for someone that wants to use eslint with prettier in a typescript project and want to fix when save a file:

settings.json:

` {
"editor.formatOnSave": true,
  "editor.formatOnType": true,
  "editor.codeActionsOnSave": {
    // "source.fixAll.eslint": "always",
    "source.organizeImports": "always"
  },
  "editor.defaultFormatter": "dbaeumer.vscode-eslint",

  /*********** ESLINT/PRETTIER ***********/
  "eslint.rules.customizations": [{ "rule": "*", "severity": "warn" }],
  "[typescript]": {
    "editor.defaultFormatter": "dbaeumer.vscode-eslint"
    // "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "eslint.codeActionsOnSave.rules": null,
  "eslint.enable": true,
  "eslint.format.enable": true,
  "prettier.printWidth": 120,
  "eslint.validate": ["javascript", "javascriptreact", "typescript", "typescriptreact"],
} `
	
.eslintrc

`{
  "parser": "@typescript-eslint/parser",
  "extends": [ "plugin:prettier/recommended", "eslint:recommended", "prettier"],
  "parserOptions": { "ecmaVersion": 2022, "sourceType": "module", "project": "tsconfig.json" },
  "plugins": ["@typescript-eslint", "prettier"],
  "overrides": [],
  "env": { "node": true },
  "rules": {
    "prettier/prettier": [
      "error",
      {
        "bracketSameLine": true,
        "useTabs": true,
        "plugins": [],
        "pluginSearchDirs": ["."],
        "semi": true,
        "trailingComma": "all",
        "singleQuote": true,
        "printWidth": 120,
        "tabWidth": 2,
        "endOfLine": "auto",
        "bracketSpacing": true,
        "arrowParens": "always"
      }
    ],
    "no-var": "error",
    "semi": "error",
    // "indent": ["error", 2, { "SwitchCase": 1 }],
    "no-multi-spaces": "error",
    "space-in-parens": "error",
    "no-multiple-empty-lines": "error",
    "prefer-const": "error",
    "prefer-template": "error",
    "no-else-return": "error",
    "no-eval": "error",
    "no-implicit-globals": "error",
    "no-implied-eval": "error",
    "no-return-assign": "error",
    "no-throw-literal": "error",
    "no-unused-expressions": "error",
    "no-useless-concat": "error",
		"no-undef": "off"
  },
  "ignorePatterns": ["node_modules/", "dist/", "swagger/", "logs/"],
  "settings": {
    "eslint.validate": ["typescript", "typescriptreact"]
  }
}`



> Ok... this is embarrassing. I commented the entire settings file, and then uncommented it line-by-line. The formatter stops working when I add this line:
> 
> ```js
> "eslint.format.enable": false, // Enabled via codeActionsOnSave (known order)
> ```
> 
> I'm a bit perplexed because I've been using this config for probably 2 months. I guess there is a tiny chance that I have never used "Format Document" (instead relying exclusively on format-on-save).
> 
> Perhaps you could advise, does this make sense (my objective is to run ESLint fixers after other formatters)?
> 
> ```js
>   "editor.codeActionsOnSave": [
>     "source.addMissingImports", // VSCode
>     "source.formatDocument", // Extension: Format Code Action
>     "source.fixAll.eslint", // Extension: ESLint
>     "source.fixAll.stylelint" // Extension: StyleLint
>   ],
>   "editor.formatOnSave": false, // Enabled via codeActionsOnSave (known order)
>   // "eslint.format.enable": false, // Do not mess with this!!
> ```
> 
> _Edit: I'm not going completely crazy -- my work project does not have this line in settings.json, and it **does** exhibit the behaviour I described earlier:_ _- fails in the same manner for the first few seconds_ _- then begins to work (seemingly at the same moment "Initializing JS/TS features" finishes)_
> 
> _...and I have now successfully replicated the same in my [first example repo](https://github.com/phits-tech/website) (by removing the `"eslint.format.enable": false` line and running `format document` soon after opening the project)_
